common/client-libs/validator-client/src/lib.rs

```diff
@@ -109,37 +87,52 @@ impl Client {
     //     let response = self.reqwest_client.get(path).send().await?.json().await?;
     // }
 
-    async fn query_validators<T>(&mut self, query: String) -> Result<T, ValidatorClientError>
+    async fn query_validators<T>(&self, query: String) -> Result<T, ValidatorClientError>
     where
         for<'a> T: Deserialize<'a>,
     {
-        // if we fail to query the first validator, push it to the back
-        let res = self.query_front_validator(query).await;
-
-        // don't bother doing any fancy validator switches if we only have 1 validator to choose from
-        if self.available_validators_rest_urls.len() > 1 {
-            if res.is_err() {
-                let front = self.available_validators_rest_urls.pop_front().unwrap();
-                self.available_validators_rest_urls.push_back(front);
-                self.failed_queries += 1;
-            }
+        let mut failed = 0;
+        let sleep_secs = 5;
+        // Randomly select a validator to query, keep querying and shuffling until we get a response
+        let mut validator_urls = self.available_validators_rest_urls().clone();
 
-            // if we exhausted all of available validators, permute the set, maybe the old ones
-            // are working again next time we try
-            if self.failed_queries == self.available_validators_rest_urls.len() {
-                self.permute_validators();
-                self.failed_queries = 0
+        // This will never exit
+        loop {
+            validator_urls.as_mut_slice().shuffle(&mut thread_rng());
+            for url in validator_urls.iter() {
+                match self.query_validator(query.clone(), &url).await {
+                    Ok(res) => return Ok(res),
+                    Err(e) => {
+                        failed += 1;
+                        error!("{}", e);
+                        error!("Total failed requests {}", failed);
+                    }
+                }
+            }
+            error!(
+                "No validators available out of {} attempted! Will try again in {} seconds. Listing all attempted:",
+                validator_urls.len(), sleep_secs
+            );
+            for url in validator_urls.iter() {
+                error!("{}", url)
             }
+            tokio::time::sleep(Duration::from_secs(sleep_secs)).await;
```

Comment:
To make it more wasm-friendly you can use the approach I've taken in gateway-client: https://github.com/nymtech/nym/blob/343c55f98136cb3a9cffe761fc66bdd6aa39cbdc/common/client-libs/gateway-client/src/client.rs#L243-L251

Reply:
I keep forgetting that this targets `wasm` as well, I'll make a script to catch this before I actually push :)

Reply:
CI already does it for you : )